### DISCUSSION TITLE
Effect function with the same functionality as Vue

### DISCUSSION THREAD

**Discussion by CrystalWindSnake** [zauberzeug/nicegui#1003] [Ideas / Feature Requests]
"I have implemented a similar `effect` function to that in Vue using some more hacky methods. Would the official support this functionality directly?"

`main.py` like this:
```python
from hack_effect import effect,checkbox,input
from nicegui import ui,Tailwind

checkbox1 = checkbox('check me')
input1 = input(label='Text', placeholder='start typing')
input2= input(label='Text', placeholder='start typing')
result = ui.label('')


# The same effect as the effect function in Vue,
# which can dynamically collect components that have changed and call this function when their values change."

'''Operation steps
1. When input1 is entered with a value of '1' and input2 is entered with a value of '95', 
    regardless of the sum result, the color of the result label remains green 
    because checkbox1 is not checked. 
2. Then, when checkbox1 is checked and input2 is entered with a value of '100', 
    the color of the result label changes to red."
'''
@effect
def test_watch():

    # After the first execution, it will monitor the changes in input1 and input2 values. 
    # This function will be re-executed. 
    # After the second execution, if the first condition (if) is passed, 
    # it will monitor the changes in input1, input2, and checkbox1 values simultaneously."
    
    if input2.value and input1.value:
        total = int(input2.value) + int(input1.value)
        result.text = str(total)

        if checkbox1.value and total>100:
            print('red')
            result.style('color:red')

        else:
            print('green')
            result.style('color:green')

ui.run()
```

`hack_effect.py` like this:
```python
from nicegui import ui
from nicegui.events import ValueChangeEventArguments
from typing import Any, Callable, Dict,List, Optional
from collections import defaultdict

g_func_id:int=None # type: ignore

# Function that collects control IDs with effect action id.
g_binds_id_mapping :dict[int,set[str]] =defaultdict(set)

# Function that maps effect action id to function object.
g_effect_func_mapping :dict[int,Callable] = {}



def run_effect_func(func:Callable):
    global g_func_id
    # start collect
    g_func_id = id(func)

    # "Clear the existing collection records."
    if g_func_id in g_binds_id_mapping:
        g_binds_id_mapping[g_func_id].clear()

    g_effect_func_mapping[g_func_id]=func

    func()

    # stop collect
    g_func_id=None # type: ignore
    

def effect(func:Callable):
    run_effect_func(func)
    
    return func
    
def collect(id:str):
    g_binds_id_mapping[g_func_id].add(id)
     

def __getattribute__(self, __name: str):
    if __name=='value':
        collect(self.id)
    return object.__getattribute__(self,__name)


def _hack_value():
    setattr(ui.checkbox,'__getattribute__',__getattribute__)
    setattr(ui.input,'__getattribute__',__getattribute__)


def all_on_changed(e:ValueChangeEventArguments):
    print('all_on_changed')
    for call_id,bind_id_set in g_binds_id_mapping.items():

        if e.sender.id in bind_id_set:
            # call effect fun
            run_effect_func(g_effect_func_mapping[call_id])


_hack_value()

def checkbox( text: str = '', *, value: bool = False):
    return ui.checkbox(text,value=value,on_change=all_on_changed)

def input(label: Optional[str] = None, *,
                 placeholder: Optional[str] = None,
                 value: str = '',
                 password: bool = False,
                 password_toggle_button: bool = False,
                 autocomplete: Optional[List[str]] = None,
                 validation: Dict[str, Callable[..., bool]] = {}):
    return ui.input(label,
                    placeholder=placeholder,
                    value=value,
                    autocomplete=autocomplete,
                    validation=validation,
                    password=password,password_toggle_button=password_toggle_button,on_change=all_on_changed)


```

**Comment by rodja** [comment]
Thanks for sharing this @CrystalWindSnake. If I understand correctly, the idea is to automatically re-run the effects-decorated function whenever a value element used inside this function changes. That is very implicit and atypical for Python. But still interesting. What do you think about making the functionality more explicit like this:

```py
@ui.on_change(checkbox1, input1, input2)
def test_watch():
    ...
```

That way we could get rid of all the auto-tracking. I renamed the decorator to `on_change` to better convey the behaviour (eg. registering `test_watch` for every provided value element in the on change handler).

I think such functionality may be useful. Especially in scenarios like your example where multiple values are evaluated to update the UI for user feedback. But Vue's `effect` function is much more general. It keeps track of reactive variables and hence also allows to re-execute when model data changes. This tracking of value elements reminds me about the discussion on a form-element we had in #801. Maybe these are related?

**Reply by CrystalWindSnake** [reply]
I think this implementation is great. I believe that allowing users to choose whether or not to use automatic monitoring of objects could lead to a better user experience, similar to the differences between React and Vue.js. Additionally, currently, component object event callback binding can only be a single function, which seems unreasonable.

**Reply by rodja** [reply]
>  think this implementation is great. I believe that allowing users to choose whether or not to use automatic monitoring of objects could lead to a better user experience

Just to make clear: are you in favour for the more explicit decorator where you need to pass the elements which should be observed or the more implicit one which you provided initially?

> currently, component object event callback binding can only be a single function, which seems unreasonable.

I totally agree. We need to allow a list of callbacks here.

**Reply by CrystalWindSnake** [reply]
I think a explicit decorator is a more secure API design. With this foundation, users can expand it in their own preferred way.

**Reply by rodja** [reply]
Great! We are on the same page. Would you like to create a pull request for this? I would really much like to make your contribution official. That way you will be part of the NiceGUI contributors list.  See [CONTRIBUTING.md](https://github.com/zauberzeug/nicegui/blob/main/CONTRIBUTING.md#pull-requests) for a short explanation.

**Reply by CrystalWindSnake** [reply]
I have read the documentation on binding model objects in nicegui, and it seems like I want this kind of functionality.

this approach seems more in line with the Vue.js style. However, my experimental code has found that the value of the variable in the effect function always gets the value of the previous moment, rather than the latest value on the interface. It seems that this is because of the mechanism of `bind_value`.

`main.py`
```python
from nicegui import ui
from hack_effect_value import effect,ref

num1 = ref('')
num2 = ref('')
checked = ref(False)

ui.checkbox('red color when checked').bind_value(checked)
ui.input().bind_value(num1)
ui.input().bind_value(num2)

result = ui.label()

# 
# total = computed(lambda: num1.value + num2.value)
# result.bind_text(total,'value')


@effect
def _():
    if num1.value and num2.value:
        print('pass')
        return

    print('no pass',num1.value,num2.value)

ui.run()
```

`hack_effect_value.py`:

```python
from typing import TypeVar, Generic, Callable
from collections import defaultdict


T = TypeVar("T")

class ref(Generic[T]):
    g_ref_id_counter : int=1

    def __init__(self,init:T) -> None:
        self.__value=init
        self.__id = ref.g_ref_id_counter

        ref.g_ref_id_counter +=1

    @property
    def value(self):
        collect(self.__id)
        return self.__value
    
    @value.setter
    def value(self, newValue:T):
        for call_id,bind_id_set in g_binds_id_mapping.items():

            if self.__id in bind_id_set and call_id:
                # call effect fun
                run_effect_func(g_effect_func_mapping[call_id])

        self.__value=newValue



g_func_id:int=None # type: ignore

# Function that collects ref IDs with effect action id.
g_binds_id_mapping :dict[int,set[int]] =defaultdict(set)

# Function that maps effect action id to function object.
g_effect_func_mapping :dict[int,Callable] = {}



def run_effect_func(func:Callable):
    global g_func_id
    # start collect
    g_func_id = id(func)

    # "Clear the existing collection records."
    if g_func_id in g_binds_id_mapping:
        g_binds_id_mapping[g_func_id].clear()

    g_effect_func_mapping[g_func_id]=func

    func()

    # stop collect
    g_func_id=None # type: ignore
    

def effect(func:Callable):
    run_effect_func(func)
    
    return func
    
def collect(id:int):
    g_binds_id_mapping[g_func_id].add(id)
```

**Reply by CrystalWindSnake** [reply]
> Great! We are on the same page. Would you like to create a pull request for this? I would really much like to make your contribution official. That way you will be part of the NiceGUI contributors list. See [CONTRIBUTING.md](https://github.com/zauberzeug/nicegui/blob/main/CONTRIBUTING.md#pull-requests) for a short explanation.

I need to first go through the underlying implementation mechanism of nicegui carefully.
